test/extensions/filters/http/cdn_loop/parser_test.cc

```diff
@@ -0,0 +1,455 @@
+  ParseContext input(value);
+  EXPECT_THAT(parseToken(input), IsOkAndHolds(ParseContext(value, 21)));
+}
+
+TEST(ParseTokenTest, TwoTokens) {
+  const std::string value = "token1 token2";
+  {
+    ParseContext input(value);
+    EXPECT_THAT(parseToken(input), IsOkAndHolds(ParseContext(value, 6)));
+  }
+  {
+    ParseContext input(value, 6);
+    EXPECT_THAT(parseToken(input), StatusIs(absl::StatusCode::kInvalidArgument));
+  }
+  {
+    ParseContext input(value, 7);
+    EXPECT_THAT(parseToken(input), IsOkAndHolds(ParseContext(value, 13)));
+  }
+}
+
+TEST(ParseTokenTest, ParseEmpty) {
+  const std::string value = "";
+  ParseContext input(value);
+  EXPECT_THAT(parseToken(input), StatusIs(absl::StatusCode::kInvalidArgument));
+}
+
+TEST(ParsePlausibleIpV6, Example) {
+  const std::string value = "[2001:DB8::1]";
+  ParseContext input(value);
+  EXPECT_THAT(parsePlausibleIpV6(input), IsOkAndHolds(ParseContext(value, 13)));
+}
+
+TEST(ParsePlausibleIpV6, ExampleLowerCase) {
+  const std::string value = "[2001:db8::1]";
+  ParseContext input(value);
+  EXPECT_THAT(parsePlausibleIpV6(input), IsOkAndHolds(ParseContext(value, 13)));
+}
+
+TEST(ParsePlausibleIpV6, ExampleIpV4) {
+  const std::string value = "[2001:db8::192.0.2.0]";
+  ParseContext input(value);
+  EXPECT_THAT(parsePlausibleIpV6(input), IsOkAndHolds(ParseContext(value, 21)));
+}
+
+TEST(ParsePlausibleIpV6, AllHexValues) {
+  const std::string value = "[1234:5678:90aA:bBcC:dDeE:fF00]";
+  ParseContext input(value);
+  EXPECT_THAT(parsePlausibleIpV6(input), IsOkAndHolds(ParseContext(value, 31)));
+}
+
+TEST(ParsePlausibleIpV6, EmptyInput) {
+  const std::string value = "";
+  ParseContext input(value);
+  EXPECT_THAT(parsePlausibleIpV6(input), StatusIs(absl::StatusCode::kInvalidArgument));
+}
+
+TEST(ParsePlausibleIpV6, BadStartDelimiter) {
+  const std::string value = "{2001:DB8::1}";
+  ParseContext input(value);
+  EXPECT_THAT(parsePlausibleIpV6(input), StatusIs(absl::StatusCode::kInvalidArgument));
+}
+
+TEST(ParsePlausibleIpV6, BadCharacter) {
+  const std::string value = "[hello]";
+  ParseContext input(value);
+  EXPECT_THAT(parsePlausibleIpV6(input), StatusIs(absl::StatusCode::kInvalidArgument));
+}
+
+TEST(ParsePlausibleIpV6, BadEndDelimiter) {
+  const std::string value = "[2001:DB8::1}";
+  ParseContext input(value);
+  EXPECT_THAT(parsePlausibleIpV6(input), StatusIs(absl::StatusCode::kInvalidArgument));
+}
+
+TEST(ParsePlausibleIpV6, EndBeforeDelimiter) {
+  const std::string value = "[2001:DB8::1";
+  ParseContext input(value);
+  EXPECT_THAT(parsePlausibleIpV6(input), StatusIs(absl::StatusCode::kInvalidArgument));
+}
+
+TEST(ParseCdnIdTest, Simple) {
+  const std::string value = "name";
+  ParseContext input(value);
+  EXPECT_THAT(parseCdnId(input), IsOkAndHolds(ParsedCdnId(ParseContext(value, 4), "name")));
+}
+
+TEST(ParseCdnIdTest, SecondInSeries) {
+  // This showcases a bug where I thought absl::string_view::substr took (start,
```

Comment:
Repharse?
// Make sure that stringpiece code correctly uses (start, end) not (start, len)
?